### DISCUSSION TITLE
Configure Network Devices via SSH (for example Mikrotik)

### DISCUSSION THREAD

**Discussion by romarioZ1000** [netbox-community/netbox#9475] [Ideas]
### NetBox version

v3.2.2

### Feature type

New functionality

### Proposed functionality

Hello, netbox community!

I have an idea to create scripts to manage network devices from web interface.

for example, here is a ready-made simple script for adding VLAN to Mikrotik.

<img width="536" alt="2022-06-03 094552" src="https://user-images.githubusercontent.com/67051805/171839994-7bde0f23-ff6f-46bf-92b1-a522c141ccde.png">

###########################################

#add netmiko to the local_requirements.txt
from django.utils.text import slugify
import paramiko
import netbox.settings
from extras.scripts import *
from netmiko import ConnectHandler
from routeros_ssh_connector import MikrotikDevice

class RunCommand(Script):
class Meta:
name = "Set VLAN on Mikrotik"
description = "Mikrotik via SSH"
field_order = [
'input_ip',
'input_bridge',
'input_vlan_name',
'input_vlan_id'
]

input_ip = StringVar(
    description="Enter the IP Address:"
)
input_bridge = StringVar(
    description="Name Bdrige:"
)

input_vlan_name = StringVar(
    description="Vlan Name:"
)

input_vlan_id = StringVar(
    description="Vlan id:"
)

def run(self, data, commit):

    command = '/interface bridge add name='+data['input_bridge'] + '\n'
    command1 = '/interface vlan add interface=ether1 name='+data['input_vlan_name'] + '()vlan-id='+data['input_vlan_id'] + '\n'
    command2 = '/interface bridge port add bridge='+data['input_bridge'] + '()interface='+data['input_vlan_name']
    command3 = '/interface bridge port add bridge='+data['input_bridge'] + '()interface=ether2 \n'
    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    client.connect(data['input_ip'], username='admin+ct80h', password='admin')
    stdin, stdout, stderr = client.exec_command(command + command1 + command2 + command3)

    mik1 = {
        "device_type": "mikrotik_routeros",
        "host": data['input_ip'],
        "username": "admin+ct",
        "password": "admin",
    }

    with ConnectHandler(**mik1) as net_connect:
        output = net_connect.send_command('/interface print \n')

        self.log_success("Mikrotik via SSH")
    return ''.join(output)

    for line in stdout:
        print(line.strip('\n'))
    client.close()

###################################################

<img width="518" alt="2022-06-03 094632" src="https://user-images.githubusercontent.com/67051805/171840159-a4de97f1-c779-4085-a300-5c8f580f6d00.png">

In this script, I enter variables myself, but I would like to enter variables from the netbox database -
for example through ObjectVar or MultiObjectVar.

############################################
    
    devices = ObjectVar(
        model=Device,
    )

    Interface_in = ObjectVar(
        model=Interface,
    )

    Interface_out = ObjectVar(
        model=Interface,
    )

    Interfacebridge = ObjectVar(
        model=Interface,
    )

    vlan_id = MultiObjectVar(
        model=VLAN,
        label='VLAN (ID)',
    )
#################################

<img width="509" alt="2022-06-03 135614" src="https://user-images.githubusercontent.com/67051805/171841203-ddc44e87-9192-4895-88be-19c2ff0ccbb3.png">

I can't pass these variables to commands that need to be written on the remote device !
'/interface bridge add name='+[bridge from Interfacebridge ] + '\n'

Can you tell me, at least with one example, how can I do this?
All scripts will be laid out in open form, so that everyone can use them!

Thanks a lot!

### Use case

Use netbox as a platform for template configuration of network devices using the netbox database.
(Mikrotik, Cisco etc)

**Comment by julianstolp** [comment]
I guess this topic should be posted in https://github.com/netbox-community/netbox/discussions/categories/ideas next time.
Anyway, inside these scripts you have complete access to the netbox database via the netboxshell (see documentation: https://docs.netbox.dev/en/stable/administration/netbox-shell/)

**Reply by kkthxbye-code** [reply]
I converted it to a discussion.

**Comment by candlerb** [comment]
> I can't pass these variables to commands that need to be written on the remote device !
'/interface bridge add name='+[bridge from Interfacebridge ] + '\n'

You haven't shown what you tried, or the complete script, or the error you got, but I'm guessing you did something like this:

```
'/interface bridge add name=' + data['Interfacebridge'] + '\n'
```

This will fail because you can't concatenate a string to an object.  You'll need to find a suitable string property, e.g.

```
'/interface bridge add name=' + data['Interfacebridge'].name + '\n'
```

Aside: Python 3.6 and later also support [string interpolation](https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings)

```
command = f'/interface bridge add name={data["Interfacebridge"].name}\n'
```

If you had used this, then actually the default `str()` conversion would have been applied automatically. [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/) is a good place to test this sort of thing interactively:

```
>>> i = Interface.objects.first()
>>> i
<Interface: en0>
>>> f'/interface bridge add name={i}\n'
'/interface bridge add name=en0\n'
>>> "foo" + i
Traceback (most recent call last):
  File "<console>", line 1, in <module>
TypeError: can only concatenate str (not "Interface") to str
>>> "foo" + i.name
'fooen0'
>>>
```

**Reply by romarioZ1000** [reply]
> > I can't pass these variables to commands that need to be written on the remote device !
> > '/interface bridge add name='+[bridge from Interfacebridge ] + '\n'
> 
> You haven't shown what you tried, or the complete script, or the error you got, but I'm guessing you did something like this:
> 
> ```
> '/interface bridge add name=' + data['Interfacebridge'] + '\n'
> ```
> 
> This will fail because you can't concatenate a string to an object. You'll need to find a suitable string property, e.g.
> 
> ```
> '/interface bridge add name=' + data['Interfacebridge'].name + '\n'
> ```
> 
> Aside: Python 3.6 and later also support [string interpolation](https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings)
> 
> ```
> command = f'/interface bridge add name={data['Interfacebridge'].name}\n'
> ```
> 
> If you had used this, then actually the default `str()` conversion would have been applied automatically. [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/) is a good place to test this sort of thing interactively:
> 
> ```
> >>> i = Interface.objects.first()
> >>> i
> <Interface: en0>
> >>> f'/interface bridge add name={i}\n'
> '/interface bridge add name=en0\n'
> >>> "foo" + i
> Traceback (most recent call last):
>   File "<console>", line 1, in <module>
> TypeError: can only concatenate str (not "Interface") to str
> >>> "foo" + i.name
> 'fooen0'
> >>>
> ```

#############################################

Good afternoon, @candlerb !!
Thanks for the advice, I tested it and it worked for me.

but can't figure out how to pass variable from 

      "Interfacebridge = ObjectVar(model=Interface,)"

 to

      "/interface bridge add name=' + data['Interfacebridge'].name + '\n' "

via 

     "i = Interface.objects.first()"   -->> how to use the filter?

##########################################################

I understand and plan to do this:

<img width="509" alt="2022-06-03 135614" src="https://user-images.githubusercontent.com/67051805/171988564-96cd6411-cfa0-4263-a803-856c706972ad.png">


the user selects parameters, for example "interface bridge", and this parameter should go to '/interface bridge add name=' + data['Interfacebridge'].name + '\n'

      "Interfacebridge = ObjectVar(model=Interface,)"

![w128h1281372336240Download](https://user-images.githubusercontent.com/67051805/171989268-dda7b654-bfa8-438e-a517-4c3939a93306.png)

      "i = Interface.objects.first()"
![w128h1281372336240Download](https://user-images.githubusercontent.com/67051805/171989268-dda7b654-bfa8-438e-a517-4c3939a93306.png)
   
      "/interface bridge add name=' + data['i'].name + '\n' "

**Reply by candlerb** [reply]
Sigh. If your script class includes

```
Interfacebridge = ObjectVar(
    model=Interface,
)
```

and your `run` function has a parameter called `data`, then the object will appear in

```
data['Interfacebridge']
```

not in `i`.  Therefore it would be:

```
command = f'/interface bridge add name={data["Interfacebridge"].name}\n'
```

EDIT: corrected. The outside quotes `f'...'` must be different from the inside quotes `["..."]`

(Aside: normally I wouldn't give the parameters names beginning with an uppercase letter. It might work, it might not. Safest to use "interfacebridge" instead of "Interfacebridge")

The problem you are likely to find next though, is that

```
interfacebridge = ObjectVar(
    model=Interface,
)
```

will produce a picklist of *all* interfaces across the whole of Netbox, not just those on the selected device.  I'm afraid I don't have a clean and simple way to populate the interface picklists dynamically. It would require some Javascript so that when the user selects a Device, it makes an AJAX call to the Netbox API to read back the interfaces available on that Device.

It will be much simpler to provide a static picklist of all interface names and bridge names you *expect* people to use.

**Reply by kkthxbye-code** [reply]
> I'm afraid I don't have a clean and simple way to populate the interface picklists dynamically. It would require some Javascript so that when the user selects a Device, it makes an AJAX call to the Netbox API to read back the interfaces available on that Device.

There's build-in support for that.

```
device = ObjectVar(
    model=Device
)
interface = ObjectVar(
    model=Interface,
    query_params={
        'device_id': '$device'
    }
)
```

Or am I misunderstanding something?

**Reply by candlerb** [reply]
Oh that's cool... I never came across that feature before :-)

https://docs.netbox.dev/en/stable/customization/custom-scripts/#objectvar

**Reply by romarioZ1000** [reply]
Thx @candlerb , @kkthxbye-code @Bierfass @bleything !

**Comment by romarioZ1000** [comment]
Thanks everyone for the advice. Here is the working config

#########################################################
```
 from django.utils.text import slugify
 import paramiko
 import netbox.settings
 from extras.scripts import *
 from netmiko import ConnectHandler
 from routeros_ssh_connector import MikrotikDevice
 from dcim.models import *
 from extras.scripts import Script, ObjectVar
 import django_filters
 from ipam.models import *
 from tenancy.models import *

   class RunCommand(Script):
      class Meta:
          name = " VLAN on Mikrotik"
          description = "Run command on Mikrotik Devices via SSH"
          field_order = [
    'device',
    'interface_in',
    'interface_out',
    'interfacebridge',
    'vlan_id',
    ]


    device = ObjectVar(
        model=Device,
    )

    interface_in = ObjectVar(
        model=Interface,
         query_params={
         'device_id': '$device'
    }
    )

    interface_out = ObjectVar(
        model=Interface,
         query_params={
         'device_id': '$device'
    }
    )

    interfacebridge = ObjectVar(
        model=Interface,
        query_params={
        'device_id': '$device'
        }
     )

    vlan_id = ObjectVar(
        model=VLAN,
        label='VLAN (ID)',
    )


    def run(self, data, commit):

        host=f'{data["device"].name}'

        command = f'/interface bridge add name=bridge_{data["interfacebridge"].name} comment=from_NB \n'
        command1 = f'/interface vlan add interface={data["interface_in"].name} name=vlan_{data["vlan_id"].name} vlan-id={data["vlan_id"].name} disable=no comment=from_NB\n'
        command2 = f'/interface bridge port add bridge=bridge_{data["interfacebridge"].name} interface={data["interface_out"].name} comment=from_NB \n'
        command3 = f'/interface bridge port add bridge={data["interfacebridge"].name} interface=vlan_{data["vlan_id"].name} comment=from_NB\n'
        client = paramiko.SSHClient()
        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        client.connect(hostname=host, username='admin+ct80h', password='admin')
        stdin, stdout, stderr = client.exec_command(command + command1 + command2 + command3)

        mik1 = {
            "device_type": "mikrotik_routeros",
            "host": host,
            "username": "admin+ct",
            "password": "admin",
        }


        with ConnectHandler(**mik1) as net_connect:
             output = net_connect.send_command('/interface print')

             self.log_success("View the information bellow")
        return ''.join(output)



        for line in stdout:
            print(line.strip('\n'))
        client.close()
```
#########################################################

**Reply by candlerb** [reply]
> Thanks everyone for the advice. Here is the working config

Thanks. Please could you edit your post to fix the formatting: add three backticks (` ``` `) on a line of their own, before and after the code.  It will make it much easier to read.

**Comment by romarioZ1000** [comment]
Hello netbox community.

I would like to continue my idea and combine my script with several types of devices (for example, Mikrotik and Cisco)

![Untitled Diagram drawio(4)](https://user-images.githubusercontent.com/67051805/172381925-1dca64df-d6b2-4a2d-a664-7a74a23ba2c6.png)

select the type of Mikrotik device, and the next line should appear where there will be 3 commands to choose from: "Show IP address", "Show User", "Show Interfaces".

But if I select the Cisco device type, then a line with two options should appear: "Show IP address", " Show User "

in this case, the desired command should be selected, because The syntax is different on different devices.

I have a simple script example with options selection.

```
from netmiko import ConnectHandler
from django.utils.text import slugify
import netbox.settings
from dcim.models import Device
from ipam.models import IPAddress
from dcim.choices import DeviceStatusChoices
from extras.scripts import *
from netaddr import *


class RunCommand(Script):
    class Meta:
        name = "View param"
        description = "View param on Mikrotik"
        field_order = ['select_device','input_command']

    select_device = ObjectVar(
        description="Select the device:",
        model = Device,
        default = None,
        required = True,
        query_params = {
             'has_primary_ip' : 'True'}
    )


    CHOICES = (
        ('ip address print', 'Show IP address'),
        ('user print', 'Show User'),
        ('interface print','Show Interfaces')
    )

    input_command = ChoiceVar(
        choices=CHOICES
    )


    def run(self, data, commit):

        address = data['select_device'].primary_ip
        input_ip = str(address.address.ip)

        mik1={
            "device_type": "mikrotik_routeros",
            "host": input_ip,
            "username": "admin+ct80h",
            "password": "admin",
             }

        cisco1 = {
            "device_type": "cisco_ios",
            "host": input_ip,
            "username": "admin",
            "password": "admin",
        }


        host=f'{data["device"].name}'
        command = data['input_command']

        with ConnectHandler(**host) as net_connect:
            output = net_connect.send_command(command)


            self.log_success("View the information bellow")
        return '' .join(output)

```


Thank you all for the tips and clarifications.

**Reply by romarioZ1000** [reply]
> Hello netbox community.
> 
> I would like to continue my idea and combine my script with several types of devices (for example, Mikrotik and Cisco)
> 
> ![Untitled Diagram drawio(4)](https://user-images.githubusercontent.com/67051805/172381925-1dca64df-d6b2-4a2d-a664-7a74a23ba2c6.png)
> 
> select the type of Mikrotik device, and the next line should appear where there will be 3 commands to choose from: "Show IP address", "Show User", "Show Interfaces".
> 
> But if I select the Cisco device type, then a line with two options should appear: "Show IP address", " Show User "
> 
> in this case, the desired command should be selected, because The syntax is different on different devices.
> 
> I have a simple script example with options selection.
> 
> ```
> from netmiko import ConnectHandler
> from django.utils.text import slugify
> import netbox.settings
> from dcim.models import Device
> from ipam.models import IPAddress
> from dcim.choices import DeviceStatusChoices
> from extras.scripts import *
> from netaddr import *
> 
> 
> class RunCommand(Script):
>     class Meta:
>         name = "View param"
>         description = "View param on Mikrotik"
>         field_order = ['select_device','input_command']
> 
>     select_device = ObjectVar(
>         description="Select the device:",
>         model = Device,
>         default = None,
>         required = True,
>         query_params = {
>              'has_primary_ip' : 'True'}
>     )
> 
> 
>     CHOICES = (
>         ('ip address print', 'Show IP address'),
>         ('user print', 'Show User'),
>         ('interface print','Show Interfaces')
>     )
> 
>     input_command = ChoiceVar(
>         choices=CHOICES
>     )
> 
> 
>     def run(self, data, commit):
> 
>         address = data['select_device'].primary_ip
>         input_ip = str(address.address.ip)
> 
>         mik1={
>             "device_type": "mikrotik_routeros",
>             "host": input_ip,
>             "username": "admin+ct80h",
>             "password": "admin",
>              }
> 
>         cisco1 = {
>             "device_type": "cisco_ios",
>             "host": input_ip,
>             "username": "admin",
>             "password": "admin",
>         }
> 
> 
>         host=f'{data["device"].name}'
>         command = data['input_command']
> 
>         with ConnectHandler(**host) as net_connect:
>             output = net_connect.send_command(command)
> 
> 
>             self.log_success("View the information bellow")
>         return '' .join(output)
> ```
> 
> Thank you all for the tips and clarifications.

or "Custom Scripts" not support this model?

**Reply by romarioZ1000** [reply]
@candlerb @kkthxbye-code @Bierfass @bleything
do you have any ideas? :)

**Reply by candlerb** [reply]
Not really.  My first reaction is that you are going to have to write some Javascript which updates the options depending on the selected device type - but I don't know if there's a clean way by which custom scripts can inject Javascript into the page.

Or maybe you can write a Plugin with a custom Model representing the actions available for a given device type, and use an [ObjectVar](https://docs.netbox.dev/en/stable/customization/custom-scripts/#objectvar) field to reference it, with suitable `query_params`.

I think you should start simple: make a ChoiceVar dropdown that covers all possible options, and if the user selects one that's not valid for the given device type, then return an error.

**Reply by julianstolp** [reply]
You could utilise the Custom links with conditional rendering and link to different scripts. The workflow would be bit different than now. You would select the devices from the deviceview and click the button that appears on the right. see https://docs.netbox.dev/en/stable/models/extras/customlink/

Example: 
Content type: DCIM > device<html>
Link text: {% if object.device_type.manufacturer.name == 'Cisco' %}Scan Switch{% endif %}
Link URL: https://netbox.mydomain.com/extras/scripts/pythonfilename.classname?deviceip={{object.primary_ip4}}

Content type: DCIM > device<html>
Link text: {% if object.device_type.manufacturer.name == 'Mikrotik' %}Scan Switch{% endif %}
Link URL: https://netbox.mydomain.com/extras/scripts/pythonfilename2.classname2?deviceip={{object.primary_ip4}}

**Reply by romarioZ1000** [reply]
Well, that's how I implemented it. I thought there was an opportunity not to create a lot of scripts, but to combine some of them.

Anyway, thanks a lot for the advice!

**Reply by candlerb** [reply]
You may be able to pass a setting to the script as a URL parameter. See #8542.

This would let you re-use the same script, just different custom links that point to it.
